internal/providers/azure/azure.go

```diff
@@ -230,3 +277,158 @@ func fetchSharedConfig(endpoint net.IP) (metadata, error) {
 
 	return config, nil
 }
+
+// fetchSshKey will fetch a single ssh key from azure and return either it or an
+// error
+//
+// The endpoint to fetch the ssh key from is stored in the GoalState under the
+// "Certificates" field. The process for how to access this endpoint and produce
+// an ssh key from the result goes as follows:
+// - generate a local rsa 2048 bit key pair
+// - provide the public key when making the request to the certificates endpoint
+// - take the contents of the "Data" field in the xml from the endpoint
+// - prepend some lines to this content to make it a valid file for openssl
+// - use the private key we generated to decrypt this value
+// - extract a public key from the decrypted contents in a pkcs8 format
+// - use ssh-keygen to convert this public key into the ssh format
+//
+// fetchSshKey can only handle a single ssh key.
+func fetchSshKey(goal GoalState) (string, error) {
+	if len(goal.RoleInstanceList) == 0 {
+		return "", fmt.Errorf("role instance list empty in goal state")
+	}
+
+	tmpDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return "", err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	keyFilePath := path.Join(tmpDir, "TransportPrivate.pem")
+	certFilePath := path.Join(tmpDir, "TransportCert.pem")
+
+	err = genCerts(keyFilePath, certFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	formattedCert, err := prepareCertForHeader(certFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	client := getClient()
+	client.Header["x-ms-cipher-name"] = []string{"DES_EDE3_CBC"}
+	client.Header["x-ms-guest-agent-public-x509-cert"] = []string{formattedCert}
+
+	certResponseBlob, err := client.Get(goal.RoleInstanceList[0].Certificates)
+	if err != nil {
+		return "", err
+	}
+
+	certResponse := certResponse{}
+	err = xml.Unmarshal(certResponseBlob, &certResponse)
+	if err != nil {
+		return "", err
+	}
+
+	encryptedSshCertFile := path.Join(tmpDir, "encrypted-ssh-cert.pem")
+	err = ioutil.WriteFile(encryptedSshCertFile, []byte(addP7mHeader(certResponse.Data, encryptedSshCertFile)), 0644)
+	if err != nil {
+		return "", err
+	}
+
+	sshCertFile := path.Join(tmpDir, "decrypted-ssh-cert.pem")
+	err = decryptFile(encryptedSshCertFile, sshCertFile, keyFilePath, certFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	sshPubkeyFile := path.Join(tmpDir, "ssh-pubkey.pem")
+	err = producePubkey(sshCertFile, sshPubkeyFile)
+	if err != nil {
+		return "", err
+	}
+
+	pubkey, err := getSshPubkey(sshPubkeyFile)
+	if err != nil {
+		return "", err
+	}
+	return pubkey, nil
+}
+
+// genCerts will generate a new 2048 bit rsa key and certificate for use in
+// certificate retrieval.
+func genCerts(keyFilePath, certFilePath string) error {
+	err := exec.ExecExpectZero("openssl", "req", "-x509", "-nodes", "-subj", "/CN=LinuxTransport", "-days", "32768", "-newkey", "rsa:2048", "-keyout", keyFilePath, "-out", certFilePath)
```

Comment:
I'm not sure if they're mandated, this command was directly copied from the waagent code: https://github.com/Azure/WALinuxAgent/blob/master/azurelinuxagent/common/utils/cryptutil.py#L30

I can play around with changing the CN and shortening the expiration, if it works with more reasonable values I'll change it to use them.